### PR TITLE
[LLVM] Use {} instead of std::nullopt to initialize empty ArrayRef

### PR DESCRIPTION
**Pull request by jayfoad**
It is almost always simpler to use {} instead of std::nullopt to
initialize an empty ArrayRef. This patch changes all occurrences I could
find in LLVM itself. In future the ArrayRef(std::nullopt_t) constructor
could be deprecated or removed.

### CODE REVIEW

**Comment by jayfoad** [comment]
@kazutakahirata what do you think?

File: llvm/lib/Target/AArch64/GISel/AArch64CallLowering.cpp [hunk 1]
```diff
@@ -1450,7 +1450,8 @@ bool AArch64CallLowering::lowerCall(MachineIRBuilder &MIRBuilder,
     if (!determineAndHandleAssignments(
             UsingReturnedArg ? ReturnedArgHandler : Handler, Assigner, InArgs,
             MIRBuilder, Info.CallConv, Info.IsVarArg,
-            UsingReturnedArg ? ArrayRef(OutArgs[0].Regs) : std::nullopt))
+            UsingReturnedArg ? ArrayRef(OutArgs[0].Regs)
+                             : ArrayRef<Register>()))
```
Comment (jayfoad) [inline llvm/lib/Target/AArch64/GISel/AArch64CallLowering.cpp:1454]: Couldn't use `{}` here due to:
```
lib/Target/AArch64/GISel/AArch64CallLowering.cpp:1453:58: error: initializer list cannot be used on the right hand side of operator ':'
```

**Review by arsenm — APPROVED** [review]

File: llvm/lib/CodeGen/SelectionDAG/SelectionDAG.cpp [hunk 2]
```diff
@@ -10428,7 +10428,7 @@ SDValue SelectionDAG::getNode(unsigned Opcode, const SDLoc &DL, SDVTList VTList,
 
 SDValue SelectionDAG::getNode(unsigned Opcode, const SDLoc &DL,
                               SDVTList VTList) {
-  return getNode(Opcode, DL, VTList, std::nullopt);
+  return getNode(Opcode, DL, VTList, ArrayRef<SDValue>());
```
Comment (jayfoad) [inline llvm/lib/CodeGen/SelectionDAG/SelectionDAG.cpp:10431]: Couldn't use `{}` here due to ambiguity between:
```
SDValue getNode(unsigned Opcode, const SDLoc &DL, SDVTList VTList, ArrayRef<SDValue> Ops);
SDValue getNode(unsigned Opcode, const SDLoc &DL, SDVTList VTList, SDValue N);
```

**Comment by kuhar** [comment]
> In future the ArrayRef(std::nullopt_t) constructor
could be deprecated or removed.

+1

**Comment by kazutakahirata** [comment]
> @kazutakahirata what do you think?

Great idea!  Thank you for cleaning this up!  The use of `std::nullopt` for `ArrayRef` is a historical artifact.  We used to have `llvm::None` for `llvm::Optional`, and we used `llvm::None` for empty `ArrayRef`, too.  It's a bad idea to use `std::nullopt` for anything other than empty `std::optional`.

**Review by kazutakahirata — APPROVED** [review]
LGTM.  Thanks!

**Pull request merged**